harvester/harvest/mit.py

```diff
@@ -68,14 +68,19 @@ def incremental_harvest_get_source_records(self) -> Iterator[Record]:
         CONFIG.check_required_env_vars()
         for zip_file_event_message in self.sqs_client.get_valid_messages_iter():
             identifier = zip_file_event_message.zip_file_identifier
-            yield Record(
-                identifier=identifier,
-                source_record=self.create_source_record_from_zip_file(
+            try:
+                source_record = self.create_source_record_from_zip_file(
                     identifier=identifier,
                     zip_file=zip_file_event_message.zip_file,
                     event=zip_file_event_message.event,
                     sqs_message=zip_file_event_message,
-                ),
+                )
+            except OSError:
+                logger.exception("File not found")
+                continue
+            yield Record(
+                identifier=identifier,
+                source_record=source_record,
```

Comment:
My knee-jerk reaction was that this is a simple and elegant solution to this issue.  And I think that's true, but I think it also short-circuits failed records handling aspects of the pipeline.

Notice here in the base `Harvester.harvest` method, where Records that encountered an error are filtered out (with the same pattern applied to each step):
```python
records = self.filter_failed_records(self.get_source_records())
```

As such, if we fail to find and parse a source record (e.g. the file is not found in S3) we could yield a `Record` instance with `Record.exception_stage="incremental_harvest_get_source_records"` and `Record.exception = <EXCEPTION OBJECT>`, we'd get the same error handling and reporting as other parts of the pipeline where something fails.

To do so, we could modify the try/except block here to still yield a `Record`, just one without a `SourceRecord` attached, and the exception encountered attached:

```python
try:
    source_record = self.create_source_record_from_zip_file(
        identifier=identifier,
        zip_file=zip_file_event_message.zip_file,
        event=zip_file_event_message.event,
        sqs_message=zip_file_event_message,
    )
    yield Record(
        identifier=identifier,
        source_record=source_record,
    )
except OSError as exc:   # <------- new logic starts here
    yield Record(
        identifier=identifier,
        source_record=None,  # <-------- Note this is None
        exception_stage="incremental_harvest_get_source_records",
        exception=exc,
    )
```

This has a couple of effects:
- the total number of records yielded from this method would actually be two (one success, one fail)
- the test `test_mit_harvester_incremental_continues_after_missing_zip_file` could then interrogate the records and assert that one has an exception (and therefore won't carry on) and one does not (will continue as normal).

Example for updates we could apply to that test:

```python
records = list(records)
assert len(records) == 2

fail_record, success_record = records
assert fail_record.identifier == "DEF456"
assert fail_record.exception_stage == "incremental_harvest_get_source_records"
assert isinstance(fail_record.exception, OSError)
assert (
    str(fail_record.exception)
    == "unable to access bucket: 'mocked_cdn_restricted' key: "
    "'cdn/geo/restricted/DEF456.zip' version: None error: An error occurred ("
    "NoSuchKey) when calling the GetObject operation: The specified key does not "
    "exist."
)
assert success_record.identifier == "SDE_DATA_AE_A8GNS_2003"
assert success_record.exception is None
```

If we lean into the failed records handling by yield a Record + exception, then we'll get the same Sentry reporting for the Record's identifier and why (to the degree that is setup and configured).

Reply:
A much better approach, updating! Thanks!

Reply:
As mentioned in the PR level comment, inside of a specific harvester it's not obvious this failure handling is happening at the orchestration level.  And, while other low level methods may utilize it, I don't think any of the `get_source_records()` methods (full and incremental) do.  So this is kind of the first for them to utilize it.